hwilib/devices/tapsigner.py

```diff
@@ -0,0 +1,408 @@
+                    continue
+                scriptcode = psbt_in.witness_script
+
+            if not is_wit:
+                if p2sh or is_p2pkh(scriptcode) or is_p2pk(scriptcode):
+                    # Add to blank_tx
+                    txin.scriptSig = scriptcode
+                # We don't know what this is, skip it
+                else:
+                    continue
+
+                # Serialize and add sighash ALL
+                ser_tx = blank_tx.serialize_without_witness()
+                ser_tx += b"\x01\x00\x00\x00"
+
+                # Hash it
+                sighash += hash256(ser_tx)
+                txin.scriptSig = b""
+            else:
+                assert psbt_in.witness_utxo is not None
+                # Calculate hashPrevouts and hashSequence
+                prevouts_preimage = b""
+                sequence_preimage = b""
+                for inputs in blank_tx.vin:
+                    prevouts_preimage += inputs.prevout.serialize()
+                    sequence_preimage += struct.pack("<I", inputs.nSequence)
+                hashPrevouts = hash256(prevouts_preimage)
+                hashSequence = hash256(sequence_preimage)
+
+                # Calculate hashOutputs
+                outputs_preimage = b""
+                for output in blank_tx.vout:
+                    outputs_preimage += output.serialize()
+                hashOutputs = hash256(outputs_preimage)
+
+                # Check if scriptcode is p2wpkh
+                if is_p2wpkh(scriptcode):
+                    _, _, wit_prog = is_witness(scriptcode)
+                    scriptcode = b"\x76\xa9\x14" + wit_prog + b"\x88\xac"
+
+                # Make sighash preimage
+                preimage = b""
+                preimage += struct.pack("<i", blank_tx.nVersion)
+                preimage += hashPrevouts
+                preimage += hashSequence
+                preimage += txin.prevout.serialize()
+                preimage += ser_string(scriptcode)
+                preimage += struct.pack("<q", psbt_in.witness_utxo.nValue)
+                preimage += struct.pack("<I", txin.nSequence)
+                preimage += hashOutputs
+                preimage += struct.pack("<I", blank_tx.nLockTime)
+                preimage += b"\x01\x00\x00\x00"
+
+                # hash it
+                sighash = hash256(preimage)
+
+            # Figure out which keypath thing is for this input
+            for pubkey, keypath in psbt_in.hd_keypaths.items():
+                if master_fp == keypath.fingerprint:
+                    # Create tuples and add to List
+                    tup = (sighash, keypath.path, i_num, pubkey)
+                    sighash_tuples.append(tup)
+
+        # Return early if nothing to do
+        if len(sighash_tuples) == 0:
+            return tx
+
+        for sighash, int_path, i_num, pubkey in sighash_tuples:
+            rec_sig = self.device.sign_digest(cvc=self.password, slot=0, digest=sighash, fullpath=path2str(int_path))
```

Comment:
What kind of exceptions will this throw? We need to be able to handle device exceptions and convert them into HWI exceptions so that the correct error codes come out. Usually this is done through decorators, see the other devices for an example.